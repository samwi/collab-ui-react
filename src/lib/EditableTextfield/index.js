import React from 'react';
import PropTypes from 'prop-types';
import { Input } from '@collab-ui/react';

/**
 * @category controls
 * @component editable-textfield
 * @variations collab-ui-react
 */

export default class EditableTextfield extends React.Component {
  static displayName = 'EditableTextfield';

  state = {
    isEditing: false,
    inputText: this.props.inputText,
  };

  componentDidUpdate = () => {
    if (this.state.isEditing && this.editText) {
      this.editText.refs.editTextfield.focus();
    }
  }

  onDoneEditing = (value) => {
    const { onDoneEditing } = this.props;

    this.setState({
      isEditing: false,
      inputText: value,
    });

    if (onDoneEditing) {
      onDoneEditing(value);
    }
  }

  editClick = () => {
    const { disabled } = this.props;

    if(disabled) {
      return;
    } else {
      this.setState({
        isEditing: true,
      });
    }
  }

  editKey = (e) => {
    const { disabled } = this.props;

    e.stopPropagation();
    e.preventDefault();

    if(disabled) {
      return;
    } else {
      this.setState({
        isEditing: true
      });
    }
  }

  onDoneKeyDown = (e) => {
    if (e.keyCode === 27) {
      this.setState({
        isEditing: false,
      });
    } else if (e.keyCode === 13) {
      this.onDoneEditing(e.target.value);
    }
  }

  render() {
    const { className } = this.props;

    if(this.state.isEditing) {
      return (
        <Input
          className={
            `cui-editable-textfield` +
            ' cui-editable-textfield__editing' +
            `${(className && ` ${className}`) || ''}`
          }
          ref={(input) => { this.editText = input; }}
          defaultValue={this.state.inputText}
          htmlId='editText'
          name='editText'
          onDoneEditing={(value) => this.onDoneEditing(value)}
          onKeyDown={(e) => this.onDoneKeyDown(e)}
        />
      );
    } else {
      return (
        <span
          role='button'
          tabIndex={0}
          className={
            `cui-editable-textfield` +
            ' cui-editable-textfield__span' +
            `${(className && ` ${className}`) || ''}`
          }
          onClick={() => this.editClick()}
          onKeyPress={(e) => this.editKey(e)}
        >
          {this.state.inputText}
        </span>
      );
    }
  }
}

EditableTextfield.propTypes = {
  /**
   * css class names
   */
  className: PropTypes.string,
  /**
   * text to be shown
   */
  inputText: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * optional disable
   */
  disabled: PropTypes.bool,
  /**
   * optional function for blur
   */
  onDoneEditing: PropTypes.func,
};

EditableTextfield.defaultProps = {
  className: '',
  inputText: '',
  disabled: false,
  onDoneEditing: null,
};


/**
* @name Default EditableTextfield
* @description default
*
* @category controls
* @component editable-textfield
* @section default
*
* @js

export default class PlaygroundComponent extends React.Component {
  consoleChange = (value) => {
    console.log(value);
  }

  render() {
    return (
      <div>
        <div className="docui-example docui-example--spacing" style={{padding: '16px'}}>
          <EditableTextfield inputText='Hello World' onDoneEditing={(value) =>this.consoleChange(value)}/>
        </div>
        <div className='cui--dark docs-example--dark' style={{padding: '16px'}}>
          <EditableTextfield inputText='Hello Dark World'/>
        </div>
    </div>
    );
  }
}

**/
