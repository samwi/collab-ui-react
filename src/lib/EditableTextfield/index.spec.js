import React from 'react';
import { shallow } from 'enzyme';
import EditableTextfield from '@collab-ui/react/EditableTextfield';

describe('tests for <EditableTextfield />', () => {
  it('should render an Editable Textfield', () => {
    const wrapper = shallow(<EditableTextfield />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render one Editable Textfield', () => {
    const container = shallow(<EditableTextfield />);

    expect(container.find('span').length).toEqual(1);
    expect(container.find('input').length).toEqual(0);
  });

  it('should click on Editable Textfield', () => {
    const container = shallow(<EditableTextfield inputText='Hello World' />);
    container.find('span').simulate('click');

    expect(container.find('span').length).toEqual(0);
    expect(container.find('Input').length).toEqual(1);
  });

  it('should click on Editable Textfield, edit and save', () => {
    const container = shallow(<EditableTextfield inputText={'Hello World'}/>);
    container.find('span').simulate('click');
    container.find('Input').simulate('keyDown', { keyCode: 27, key: 'Enter' });

    expect(container.find('span').text()).toEqual('Hello World');
    expect(container.find('span').length).toEqual(1);
    expect(container.find('Input').length).toEqual(0);
  });
});
