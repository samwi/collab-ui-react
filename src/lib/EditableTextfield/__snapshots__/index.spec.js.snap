// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <EditableTextfield /> should render an Editable Textfield 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditableTextfield
    className=""
    disabled={false}
    inputText=""
    onDoneEditing={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "",
      "className": "cui-editable-textfield cui-editable-textfield__span",
      "onClick": [Function],
      "onKeyPress": [Function],
      "role": "button",
      "tabIndex": 0,
    },
    "ref": null,
    "rendered": "",
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "",
        "className": "cui-editable-textfield cui-editable-textfield__span",
        "onClick": [Function],
        "onKeyPress": [Function],
        "role": "button",
        "tabIndex": 0,
      },
      "ref": null,
      "rendered": "",
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
